### PR TITLE
[SPARK-49962][SQL] Simplify AbstractStringTypes class hierarchy

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?
Simplifying the AbstractStringType hierarchy.

### Why are the changes needed?
The addition of trim-sensitive collation (#48336) highlighted the complexity of extending the existing AbstractStringType structure. Besides adding a new parameter to all types inheriting from AbstractStringType, it caused changing the logic of every subclass as well as changing the name of a derived class StringTypeAnyCollation into StringTypeWithCaseAccentSensitivity which could again be subject to change if we keep adding new specifiers.

Looking ahead, the introduction of support for indeterminate collation would further complicate these types. To address this, the proposed changes simplify the design by consolidating common logic into a single base class. This base class will handle core functionality such as trim or indeterminate collation, while a derived class, StringTypeWithCollation (previously awkwardly called StringTypeWithCaseAccentSensitivity), will manage collation specifiers.

This approach allows for easier future extensions: fundamental checks can be handled in the base class, while any new specifiers can be added as optional fields in StringTypeWithCollation.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
With existing tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by stefankandic** [comment]
@uros-db @jovanpavl-db Please take a look and let me know if this makes sense

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 1]
```diff
@@ -21,25 +21,35 @@ import org.apache.spark.sql.internal.SqlApiConf
...
+      case st: StringType =>
+        canUseTrimCollation(st) && acceptsStringType(st)
+      case _ =>
+        false
+    }
+  }
+  def acceptsStringType(other: StringType): Boolean
+
+  private[sql] def canUseTrimCollation(other: StringType): Boolean =
+    supportsTrimCollation || !other.usesTrimCollation
```
Comment (uros-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:42]:
```suggestion

  private[sql] def canUseTrimCollation(other: StringType): Boolean =
    supportsTrimCollation || !other.usesTrimCollation

  def acceptsStringType(other: StringType): Boolean
```

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 2]
```diff
@@ -21,25 +21,35 @@ import org.apache.spark.sql.internal.SqlApiConf
...
-    supportsTrimCollation || !other.asInstanceOf[StringType].usesTrimCollation
+
+  override private[sql] def acceptsType(other: DataType): Boolean = {
+    other match {
+      case st: StringType =>
+        canUseTrimCollation(st) && acceptsStringType(st)
+      case _ =>
+        false
+    }
+  }
```
Comment (uros-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:38]:
```suggestion
  override private[sql] def acceptsType(other: DataType): Boolean = other match {
      case st: StringType => canUseTrimCollation(st) && acceptsStringType(st)
      case _ => false
    }
```

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala [hunk 3]
```diff
@@ -21,25 +21,35 @@ import org.apache.spark.sql.internal.SqlApiConf
 import org.apache.spark.sql.types.{AbstractDataType, DataType, StringType}
 
 /**
- * AbstractStringType is an abstract class for StringType with collation support. As every type of
- * collation can support trim specifier this class is parametrized with it.
+ * AbstractStringType is an abstract class for StringType with collation support.
  */
-abstract class AbstractStringType(private[sql] val supportsTrimCollation: Boolean = false)
+abstract class AbstractStringType(supportsTrimCollation: Boolean = false)
```
Comment (uros-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:26]: why did you remove `private[sql] val`? let's keep it private unless there's a need

**Review by uros-db — APPROVED** [review]
thanks for the changes, left a couple of comments

otherwise lgtm, let's go for it

**Comment by uros-db** [comment]
also, let's just mention in the PR description that we're also doing some renaming (which is most of the code changes in this PR I'd say)

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala (hunk 3 above)
Comment (stefankandic) [reply sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:26]: Why does it need to private? Looks like other classes inheriting from AbstractDataType like AbstractMapType don't have their members private either.

**Review by jovanpavl-db — APPROVED** [review]
Looks great, thanks!

File: sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala (hunk 3 above)
Comment (uros-db) [reply sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:26]: can we keep it private to SQL in all of them? no need to expose stuff unless there's a need to expose stuff

Comment (HyukjinKwon) [reply sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:26]: `sql/internal` is already a private package so we won;t need to mark it as `private[sql]`.   We should add a `package.scala` and document so, see https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/package.scala

Comment (HyukjinKwon) [reply sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala:26]: See also https://issues.apache.org/jira/browse/SPARK-16964 and https://issues.apache.org/jira/browse/SPARK-16813

**Review by MaxGekk — APPROVED** [review]
